Unclear error message thrown when using unsupported JOIN type as a second JOIN in SQL expression

When a user executes the following query, when there are two joins and the second one is unsupported:
```
SELECT person.name, work.name, work.address, pojo.bigIntVal
FROM pojo_map pojo 
LEFT JOIN person_map person ON pojo.__key = person.id 
FULL OUTER JOIN work_map work ON pojo.__key = work.id 
WHERE work.name = 'SCHOOL'
```
The following stacktrace is being thrown. That looks rather like a bug, and not like an unsupported clause.
```
com.hazelcast.sql.HazelcastSqlException
	at com.hazelcast.sql.impl.QueryUtils.toPublicException(QueryUtils.java:76)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:233)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:199)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:195)
	at com.hazelcast.sql.SqlService.execute(SqlService.java:146)
	at com.hazelcast.jet.sql.impl.connector.map.manual_testing.ManualTest.test1(ManualTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: com.hazelcast.sql.impl.QueryException
	at com.hazelcast.sql.impl.QueryException.error(QueryException.java:59)
	at com.hazelcast.sql.impl.calcite.parse.QueryParser.parse(QueryParser.java:90)
	at com.hazelcast.sql.impl.calcite.OptimizerContext.parse(OptimizerContext.java:137)
	at com.hazelcast.sql.impl.calcite.CalciteSqlOptimizer.prepare(CalciteSqlOptimizer.java:127)
	at com.hazelcast.sql.impl.SqlServiceImpl.prepare(SqlServiceImpl.java:287)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:268)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:220)
	... 29 more
Caused by: java.lang.NullPointerException
	at com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeUtils.isNumericIntegerType(HazelcastTypeUtils.java:198)
	at com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeFactory.createTypeWithNullability(HazelcastTypeFactory.java:152)
	at org.apache.calcite.sql.validate.JoinNamespace.validateImpl(JoinNamespace.java:55)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:84)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateFrom(SqlValidatorImpl.java:3262)
	at com.hazelcast.jet.sql.impl.validate.JetSqlValidator.validateFrom(JetSqlValidator.java:100)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3510)
	at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:60)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:84)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1110)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1084)
	at com.hazelcast.sql.impl.calcite.validate.HazelcastSqlValidator.validateQuery(HazelcastSqlValidator.java:165)
	at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:232)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1059)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:766)
	at com.hazelcast.jet.sql.impl.validate.JetSqlValidator.validate(JetSqlValidator.java:83)
	at com.hazelcast.sql.impl.calcite.parse.QueryParser.parse(QueryParser.java:103)
	at com.hazelcast.sql.impl.calcite.parse.QueryParser.parse(QueryParser.java:77)
	... 34 more
```

**Expected behavior**
Error message of the following format should be thrown:
```
com.hazelcast.sql.HazelcastSqlException: From line 1, column 91 to line 1, column 94: FULL join not supported
```

--- a/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeFactory.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeFactory.java
@@ -237,6 +237,8 @@ public final class HazelcastTypeFactory extends SqlTypeFactoryImpl {
 
     @Override
     public RelDataType createJoinType(RelDataType... types) {
+        // Calcite's implementation of createJoinType() returns RelCrossType.
+        // RelCrossType.getSqlTypeName() returns null. We assume in several places that it's non null...
         return createUnknownType();
     }
 }

--- a/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeFactory.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeFactory.java
@@ -234,4 +234,9 @@ public final class HazelcastTypeFactory extends SqlTypeFactoryImpl {
 
         return HazelcastIntegerType.create((HazelcastIntegerType) maxBitWidthType, targetType.isNullable());
     }
+
+    @Override
+    public RelDataType createJoinType(RelDataType... types) {
+        return createUnknownType();
+    }
 }

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnsupportedFeaturesTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlUnsupportedFeaturesTest.java
@@ -61,6 +61,21 @@ public class SqlUnsupportedFeaturesTest extends SqlTestSupport {
     }
 
     @Test
+    public void test_multiFullJoin() {
+        TestBatchSqlConnector.create(sqlService, "b", 0);
+
+        assertThatThrownBy(() -> sqlService.execute(
+                    "SELECT 1 FROM b AS b1 JOIN b AS b2 ON b1.v = b2.v FULL OUTER JOIN b AS b3 ON b2.v = b3.v"))
+                .hasCauseInstanceOf(QueryException.class)
+                .hasMessageContaining("FULL join not supported");
+
+        assertThatThrownBy(() -> sqlService.execute(
+                    "SELECT 1 FROM b AS b1 JOIN b AS b2 ON b1.v = b2.v FULL JOIN b AS b3 ON b2.v = b3.v"))
+                .hasCauseInstanceOf(QueryException.class)
+                .hasMessageContaining("FULL join not supported");
+    }
+
+    @Test
     public void test_semiJoin() {
         TestBatchSqlConnector.create(sqlService, "b", 0);
 

